### PR TITLE
Aes alert and reset test

### PR DESCRIPTION
**Pull request by rasmus-madsen**
This PR contains a new alert/reset test
that will either flip a bit in a onehot state machine and verify that:

- the fatal alert gets set
- that we cannot get information (like the state of the core)
- that the module can recover on a reset

or it will just pull reset at a random time and verify that we can recover.

### CODE REVIEW

File: hw/ip/aes/dv/env/aes_scoreboard.sv [hunk 1]
```diff
@@ -440,46 +454,45 @@ class aes_scoreboard extends cip_base_scoreboard #(
 
   // takes items from the item queue and builds full
   // aes_messages with both input data and output data.
-  task rebuild_message();
+virtual  task rebuild_message();
```
Comment (cindychip) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:457]: super nit: you have one more space here :)

File: hw/ip/aes/dv/env/aes_scoreboard.sv [hunk 2]
```diff
@@ -440,46 +454,45 @@ class aes_scoreboard extends cip_base_scoreboard #(
...
+              `uvm_info(`gfn, $sformatf("\t ----| got item from item fifo "), UVM_MEDIUM)
               if (full_item.message_start() || (full_item.start_item && full_item.manual_op)) begin
                 `downcast(msg_clone, message.clone());
                 msg_fifo.put(msg_clone);
                 message = new();
                 if (full_item.start_item && full_item.manual_op) begin
-                   message.skip_msg = 1;
-                  `uvm_info(`gfn, $sformatf("\n setting skip msg"), UVM_MEDIUM)
+                  // only set skip if this is not a real start message
+                   message.skip_msg = ~full_item.message_start();
```
Comment (cindychip) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:495]: sorry also super nit stuff: looks like you have an extra space at the beginning of this line.

File: hw/ip/aes/dv/env/seq_lib/aes_stress_vseq.sv [hunk 3]
```diff
@@ -7,15 +7,17 @@ class aes_stress_vseq extends aes_base_vseq;
   `uvm_object_utils(aes_stress_vseq)
 
   `uvm_object_new
-  aes_message_item my_message;
+    aes_message_item my_message;
```
Comment (cindychip) [inline hw/ip/aes/dv/env/seq_lib/aes_stress_vseq.sv:10]: super nit: extra space

File: hw/ip/aes/dv/tests/aes_base_test.sv [hunk 4]
```diff
@@ -16,6 +16,12 @@ class aes_base_test extends cip_base_test #(
 
   //this will serve as the default setting.
   // overrides should happen in the specific testcase.
+
+  virtual function void end_of_elaboration_phase(uvm_phase phase);
+    super.end_of_elaboration_phase(phase);
+   // uvm_top.print_topology();
+  endfunction // end_of_elaboration
```
Comment (cindychip) [inline hw/ip/aes/dv/tests/aes_base_test.sv:23]: Is this change intended to be here?
Maybe you can set a verbosity to print out the topology?

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          error: begin
+            cfg.clk_rst_vif.wait_clks(cfg.inj_delay);
+            if (cfg.error_types.mal_inject && (cfg.flip_rst == Flip_bits)) begin
+              randomize(mal_error)  with { $countones(mal_error) > 1; };
+              if (!uvm_hdl_check_path("tb.dut.u_aes_core.u_ctrl_reg_shadowed.committed_q")) begin
+                `uvm_fatal(`gfn, $sformatf("\n\t ----| PATH NOT FOUND"))
+              end else begin
+                $assertoff(0, "tb.dut.u_aes_core.AesModeValid");
+                $assertoff(0, "tb.dut.u_aes_core.u_aes_control.AesModeValid");
+                uvm_hdl_force("tb.dut.u_aes_core.u_ctrl_reg_shadowed.committed_q[6:1]", mal_error);
```
Comment (cindychip) [inline hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: Since the alert checking is disabled in scb, do you think is worth to check alert ping in sequence here? I think maybe you can try some logic similar to here: https://github.com/lowRISC/opentitan/blob/master/hw/ip/otp_ctrl/dv/env/seq_lib/otp_ctrl_background_chks_vseq.sv#L55
1. Wait for alert to trigger.
2. Then check the handshake complete, if fatal alert, check the alert always triggers.
Just suggestions, not sure if it is suitable for AES here.

Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @cindychip 
the aes_base_sequence and scoreboard is already reading the output of the status register to check the fatal bit gets set.
but I guess this could be a good check to add in the case that the register gets set but the actual alert isn't raised.

File: hw/ip/aes/dv/env/aes_scoreboard.sv (hunk 1 above)
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_scoreboard.sv:457]: fixed

File: hw/ip/aes/dv/env/aes_scoreboard.sv (hunk 2 above)
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_scoreboard.sv:495]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_stress_vseq.sv (hunk 3 above)
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_stress_vseq.sv:10]: fixed

File: hw/ip/aes/dv/tests/aes_base_test.sv (hunk 4 above)
Comment (rasmus-madsen) [reply hw/ip/aes/dv/tests/aes_base_test.sv:23]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv (hunk 5 above)
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: I added this check where the fatal is detected which seem more intuitive in this case

**Review by cindychip — APPROVED** [review]
LGTM
